src/backends/gpu_kernels.hpp

```diff
@@ -0,0 +1,434 @@
+#pragma once
+
+#include <cstdint>
+#include <cfloat>
+
+#include <iostream>
+#include <limits>
+#include <string>
+#include <vector>
+
+namespace nest {
+namespace mc {
+namespace gpu {
+
+namespace impl {
+    //
+    // Compile time constants for block-interleave matrix storage
+    //
+
+    // Number of matrices per block in block-interleaved storage
+    __host__ __device__
+    constexpr inline int block_dim() {
+        return 32;
+    }
+
+    // The number of threads per matrix in the interleave and reverse-interleave
+    // operations.
+    __host__ __device__
+    constexpr inline int load_width() {
+        return 32;
+    }
+
+    // The alignment of matrices inside the block-interleaved storage.
+    __host__ __device__
+    constexpr inline int matrix_padding() {
+        return load_width();
+    }
+
+    // Number of threads per warp
+    // This has always been 32, however it may change in future NVIDIA gpus
+    __host__ __device__
+    constexpr inline int threads_per_warp() {
+        return 32;
+    }
+
+    // The minimum number of bins required to store n values where the bins have
+    // dimension of block_size.
+    __host__ __device__
+    constexpr inline int block_count(int n, int block_size) {
+        return (n+block_size-1)/block_size;
+    }
+
+    // The smallest size of a buffer required to store n items in such that the
+    // buffer has size that is a multiple of block_dim.
+    inline int padded_size (int n, int block_dim) {
```

Comment:
I had it that way to avoid computing `n%block_dim` twice in the same expression. Not for performance reasons: just because it makes it easier to understand the code (in my opinion).
I changed it to `constexpr`... maybe when we move to C++14 we won't need to have this discussion.